conjure-go-contract/errors/http.go

```diff
@@ -15,40 +15,30 @@
 package errors
 
 import (
+	"encoding/json"
 	"net/http"
 
 	"github.com/palantir/conjure-go-runtime/conjure-go-contract/codecs"
 )
 
-// ErrorFromResponse extract serializable error from the given response.
-//
-// TODO This function is subject to change.
-func ErrorFromResponse(response *http.Response) (SerializableError, error) {
-	var unmarshalled SerializableError
-	if err := codecs.JSON.Decode(response.Body, &unmarshalled); err != nil {
-		return SerializableError{}, err
-	}
-	return unmarshalled, nil
-}
-
 // WriteErrorResponse writes error to the response writer.
 //
 // TODO This function is subject to change.
-func WriteErrorResponse(w http.ResponseWriter, e SerializableError) {
-	marshalledError, err := codecs.JSON.Marshal(e)
-	if err != nil {
-		// Falling back to marshalling error without parameters.
-		// This should always succeed given.
-		marshalledError, _ = codecs.JSON.Marshal(
-			SerializableError{
-				ErrorCode:       e.ErrorCode,
-				ErrorName:       e.ErrorName,
-				ErrorInstanceID: e.ErrorInstanceID,
-				Parameters:      nil,
-			},
-		)
+func WriteErrorResponse(w http.ResponseWriter, e Error) {
+	var marshalledError []byte
+	var err error
+
+	// First try to marshal with custom handling (if present)
+	if marshaler, ok := e.(json.Marshaler); ok {
+		marshalledError, err = codecs.JSON.Marshal(marshaler)
 	}
+	// If we fail, use best-effort conversion to SerializableError.
+	if marshalledError == nil || err != nil {
+		// serializeError() handles param failures, so this should never fail
+		marshalledError, _ = codecs.JSON.Marshal(serializeError(e))
```

Comment:
Isn't it safer to handle this error? If it throws, `marshalledError` will be nil. Not sure if that's a problem on `w.Write`.

Reply:
Yeah, I fought with this for a long time. Maybe it's more understandable if I just bring `serializeError` into this file/function. I thought it was going to get used in more than one place but ended up being wrong.

The idea (which that helper function handles) is that if we fail to marshal it's going to be due to the params and not the primitive fields. So we null out the params on errors and just assume marshalling the other three fields will work.

Reply:
I looked at `w.Write` and it looks like it should gracefully handle a `nil` for `marshalledError`, so I guess this is ok. Worst case, you don't get anything in the response body.